docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,304 @@
+
+There are 2 methods described for informing the cluster autoscaler about the resource needs of the
+nodes in each node group: through a status field on Infrastructure Machine Templates, and through
+annotations on MachineSets or MachineDeployments. The first method requires updates to a infrastructure provider's
+controllers and will require more coordination between developers and users. The second method
+requires less direct intervention from infrastructure providers and puts more resposibility on users, for
+this additional responsibility the users gain immediate access to the feature. These methods are
+mutually exclusive, and the annotations will take preference when specified.
+
+#### Infrastructure Machine Template Status Updates
+
+Infrastructure providers should add a field to the `status` of any Infrastructure Machine Template they reconcile.
+This field will contain the CPU, memory, and GPU resources associated with the machine described by
+the template.  Internally, this field will be represented by a Go `map` type utilizing named constants
+for the keys and `k8s.io/apimachinery/pkg/api/resource.Quantity` as the values (similar to how resource
+limits and requests are handled for pods).
+
+It is worth mentioning that the Infrastructure Machine Templates are not usually reconciled by themselves.
+Each infrastructure provider will be responsible for determining the best implementation for adding the
+status field based on the information available on their platform.
+
+**Example implementation in Docker provider**
+```
+// these constants will be carried in Cluster API, but are repeated here for clarity
+const (
+    AutoscalerResourceCPU ResourceName = "cpu"
+    AutoscalerResourceMemory ResourceName = "memory"
+    AutoscalerResourceGPU ResourceName = "gpu"
+)
+
+// AutoscalerMachineStatus defines information about individual machines used by the cluster autoscaler to perform scale from zero operations
+type AutoscalerMachineStatus struct {
+    Capacity ResourceList `json:"capacity,omitempty"`
+}
+
+// AutoscalerMachineTemplateStatus defines information that is used by the cluster autoscaler to perform scale from zero operations
+type AutoscalerMachineTemplateStatus struct {
+    Machine AutoscalerMachineStatus `json:"machine,omitempty"`
+}
+
+// DockerMachineTemplateStatus defines the observed state of a DockerMachineTemplate
+type DockerMachineTemplateStatus struct {
+    Autoscaler AutoscalerMachineTemplateStatus `json:"autoscaler,omitempty"`
+}
+
+// DockerMachineTemplate is the Schema for the dockermachinetemplates API.
+type DockerMachineTemplate struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec DockerMachineTemplateSpec     `json:"spec,omitempty"`
+    Status DockerMachineTemplateStatus `json:"status,omitempty"`
+}
+```
+_Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/kubernetes/pkg/apis/core/types`_
```

Comment:
```suggestion
_Note: the `ResourceList` and `ResourceName` referenced are from k8s.io/api/core/v1`_
```
nit: we probably don't want to import k/k

Reply:
> nit: we probably don't want to import k/k

completely agree!